
From: Jim Houston <jim.houston@comcast.net>

- Add IDR_INIT() DEFINE_IDR() constructors.  Often idr's are singletons
  and having to cook up an initcall for them is a pain.


---

 25-akpm/include/linux/idr.h |   12 ++
 2 files changed, 128 insertions(+), 76 deletions(-)

diff -puN include/linux/idr.h~idr-extra-features include/linux/idr.h
--- 25/include/linux/idr.h~idr-extra-features	Fri Mar 12 14:32:34 2004
+++ 25-akpm/include/linux/idr.h	Fri Mar 12 14:32:34 2004
@@ -23,6 +23,7 @@
 # error "BITS_PER_LONG is not 32 or 64"
 #endif
 
+#define IDR_SIZE (1 << IDR_BITS)
 #define IDR_MASK ((1 << IDR_BITS)-1)
 
 /* Define the size of the id's */
@@ -53,6 +54,17 @@ struct idr {
 	spinlock_t	  lock;
 };
 
+#define IDR_INIT(name)	\
+{								\
+	.top		= NULL,					\
+	.id_free	= NULL,					\
+	.count		= 0,					\
+	.layers 	= 0,					\
+	.id_free_cnt	= 0,					\
+	.lock		= SPIN_LOCK_UNLOCKED,			\
+}
+#define DEFINE_IDR(name)	struct idr name = IDR_INIT(name)
+
 /*
  * This is what we export.
  */
